### PR TITLE
Redis cluster support

### PR DESCRIPTION
**Pull request by dustinrue**
This PR adds support for Redis running in a cluster mode. This is common when using AWS ElastiCache.

By default, the code will connect to Redis as it did previously. If W3TC_REDIS_CLUSTER is defined and set to true the connection will be made using RedisCluster instead of Redis. When used with AWS ElastiCache, only the Configuration Endpoint is needed.

### CODE REVIEW

File: w3-total-cache.php [hunk 1]
```diff
@@ -68,6 +68,11 @@
 	if ( defined( 'WP_CLI' ) && WP_CLI ) {
 		require_once W3TC_DIR . '/Cli.php';
 	}
+	
+	// Set W3TC_REDIS_CLUSTER to true if using Redis in cluster mode
```
Comment (cssjoe) [inline w3-total-cache.php:72]: This new code block should go in the file `w3-total-cache-api.php`.

File: Cache_Redis.php [hunk 2]
```diff
@@ -344,35 +344,42 @@ private function _get_accessor( $key ) {
...
-				$accessor = new \Redis();
-
-				if ( substr( $server, 0, 5 ) == 'unix:' ) {
-					if ( $this->_persistent ) {
-						$accessor->pconnect( trim( substr( $server, 5 ) ),
-							null, null, $this->_instance_id . '_' . $this->_dbid );
+				if ( W3TC_REDIS_CLUSTER === true ) {
+					$accessor = new \RedisCluster(NULL, $this->_servers, 1.5, 1.5, $this->_persistent, $this->_password);
+				}
+				else {
```
Comment (cssjoe) [inline Cache_Redis.php:350]: Formatting is a little off.  Please change to:
`} else {`

File: Cache_Redis.php [hunk 3]
```diff
@@ -344,35 +344,42 @@ private function _get_accessor( $key ) {
...
 		else {
 			try {
-				$server = $this->_servers[$index];
-				$accessor = new \Redis();
-
-				if ( substr( $server, 0, 5 ) == 'unix:' ) {
-					if ( $this->_persistent ) {
-						$accessor->pconnect( trim( substr( $server, 5 ) ),
-							null, null, $this->_instance_id . '_' . $this->_dbid );
+				if ( W3TC_REDIS_CLUSTER === true ) {
```
Comment (cssjoe) [inline Cache_Redis.php:347]: A better check would be:
`if ( defined( 'W3TC_REDIS_CLUSTER' ) && W3TC_REDIS_CLUSTER ) {`

File: Cache_Redis.php [hunk 4]
```diff
@@ -331,7 +331,7 @@ public function counter_get( $key ) {
 	}
 
 	private function _get_accessor( $key ) {
-		if ( count( $this->_servers ) <= 1 )
+		if ( count( $this->_servers ) <= 1 || W3TC_REDIS_CLUSTER )
```
Comment (cssjoe) [inline Cache_Redis.php:334]: I better condition would be:
`if ( count( $this->_servers ) <= 1 || ( defined( 'W3TC_REDIS_CLUSTER' ) && W3TC_REDIS_CLUSTER ) ) {`

File: Generic_AdminActions_Test.php [hunk 5]
```diff
@@ -47,25 +47,52 @@ function w3tc_test_redis() {
...
 			$success = true;
 
-			foreach ( $servers as $server ) {
-				@$cache = Cache::instance( 'redis', array(
-						'servers' => $server,
-						'persistent' => false,
-						'password' => $password,
-						'dbid' => $dbid
-					) );
+			if ( W3TC_REDIS_CLUSTER === true ) {
```
Comment (cssjoe) [inline Generic_AdminActions_Test.php:50]: I better check would be:
`if ( defined( 'W3TC_REDIS_CLUSTER' ) && W3TC_REDIS_CLUSTER ) {`

**Review by cssjoe — CHANGES_REQUESTED** [review]
@dustinrue Thank you for submitting this contribution.  There are a few changes requested.  I also requested another one of our developers to review the request.  Please review my comments and make the changes.  Thank you.

**Comment by dustinrue** [comment]
Thank you for the feedback. I have made the requested changes.

**Comment by maxicus** [comment]
there are 3 redis connectivity classes available at the same time (Redis, RedisArray, RedisCluster). 
to keep similar with other parts of the code it should be controlled by string config parameter instead of hidden define, similar to `memcached.binary_protocol`

**Comment by dustinrue** [comment]
I made some updates but I'm waiting on someone to help me test the changes

**Review by cssjoe — CHANGES_REQUESTED** [review]
See Max's previous comment:
>there are 3 redis connectivity classes available at the same time (Redis, RedisArray, RedisCluster).
to keep similar with other parts of the code it should be controlled by string config parameter instead of hidden define, similar to memcached.binary_protocol

Also, there are merge conflicts that have to be resolved.  There have been many updates since this pull request was opened.  Please resolve all merge conflicts and make updates based on Max's comment above.  We'll be happy to review this again after it has been updated.  Thank you.

**Comment by cssjoe** [comment]
This pull request needs changes and have merge conflicts resolved.

**Pull request closed without merging**